src/rgw/rgw_asio_frontend.cc

```diff
@@ -1075,12 +1075,31 @@ void AsioFrontend::stop()
   ldout(ctx(), 4) << "frontend initiating shutdown..." << dendl;
 
   going_down = true;
+  const bool graceful_stop{ g_ceph_context->_conf->rgw_graceful_stop };
+  if (graceful_stop) {
+    // reject new connections immediately
+    boost::system::error_code ec;
+    for (auto& l : listeners) {
+      l.acceptor.close(ec);
+    }
```

Comment:
this part is duplicated just below, which shouldn't be necessary. can you just add a single `if (graceful_stop) {` block under that which only calls lock/unlock?

Reply:
@cbodley ack, thank you, ultimately the functionality can be distilled to adding a `pause_mutex` lock/unlock at the correct place, please check amend